yktrace/src/hwt/mapper.rs

```diff
@@ -134,59 +130,66 @@ impl<'a> HWTMapper<'a> {
+    ///    every raw trace will start by returning out of the foreign code that turned on trace
+    ///    collection. We omit this prefix.
+    ///
+    ///  - To save memory, in the returned trace consecutive unmappable block entries will have
+    ///    been collapsed into one. An unknown block entry therefore means "one or more unmappable
+    ///    blocks".
+    pub fn map_trace(
         &mut self,
-        last_yielded: Option<&IRBlock>,
-    ) -> Option<Result<IRBlock, HWTracerError>> {
-        while self.buffered_irblocks.is_empty() {
-            let block = self.trace_iter.next();
-            if block.is_none() {
-                return None;
-            }
-            let block = block.unwrap();
-            if let Err(e) = block {
-                return Some(Err(e));
-            }
-            let irblocks = self.map_block(block.as_ref().unwrap());
+        mut trace_iter: &'a mut dyn Iterator<Item = Result<Block, HWTracerError>>,
+    ) -> Result<Vec<IRBlock>, HWTracerError> {
+        let mut ret: Vec<IRBlock> = Vec::new();
+
+        for block in &mut trace_iter {
+            let irblocks = self.map_block(&block?);
             if irblocks.is_empty() {
                 // The block is unmappable. Insert a IRBlock that indicates this, but only if the
-                // trace isn't empty. We also take care to collapse repeated unmappable blocks into
-                // a single unmappable IRBlock.
-                if let Some(blk) = last_yielded {
-                    if !blk.is_unmappable() {
-                        self.buffered_irblocks.push(IRBlock::unmappable());
-                    }
+                // trace isn't empty (we never report the leading unmappable code in a trace). We
+                // also take care to collapse consecutive unmappable blocks into one.
+                if let Some(last) = ret.last_mut() {
+                    if !last.is_unmappable() {
+                        ret.push(IRBlock::unmappable());
+                    } // otherwise don't push, thus collapsing repeated unmappable blocks.
                 }
             } else {
                 for irblock in irblocks.into_iter() {
-                    if let Some(irblock) = irblock {
-                        // The `BlockDisambiguate` pass in ykllvm ensures that no high-level LLVM
-                        // IR block ever branches straight back to itself, so if we see the same
-                        // high-level block more than once consecutively in a trace, then we know
-                        // that the IR block has been lowered to multiple machine blocks during
-                        // code-gen, and that we should only push the IR block once.
-                        if last_yielded != Some(&irblock) {
-                            self.buffered_irblocks.push(irblock);
+                    if let Some(irb) = irblock {
+                        if let Some(last) = ret.last() {
```

Comment:
Maybe a `match` for this to bring out the control flow?

```rust
match ret.last() {
  Some(last) if &irb != last { ... },
  Some(_) => {// the big comment},
  None => {ret.push(irb)
}
```

Reply:
And bring it out we shall: c4f521c4c72fbc69c2fd593776c254ffeff306cf